src/router.ts

```diff
@@ -0,0 +1,125 @@
+  name: Name;
+  description?: string;
+  request: z.ZodType<any, any, any>;
+  response: z.ZodType<any, any, any>;
+};
+
+type Method = 'GET' | 'DELETE' | 'PUT' | 'POST' | 'PATCH';
+type RoughRoute = `${Method} ${string}`;
+
+type ZodSchema = {
+  [route: string]: RouterEndpointDefinition<string>;
+};
+
+export type OneSchemaRouterConfig<R extends Router<any, any>> = {
+  using: R;
+  introspection: IntrospectionConfig | undefined;
+};
+
+export class OneSchemaRouter<
+  Schema extends ZodSchema,
+  R extends Router<any, any>,
+> {
+  private router: R;
+
+  private constructor(
+    private schema: Schema,
+    { introspection, using: router }: OneSchemaRouterConfig<R>,
+  ) {
+    this.router = router;
+    if (introspection) {
+      router.get(introspection.route, (ctx, next) => {
+        const response: IntrospectionResponse = {
+          serviceVersion: introspection.serviceVersion,
+          schema: convertRouterSchemaToJSONSchemaStyle(this.schema),
+        };
+
+        ctx.body = response;
+        ctx.status = 200;
+        return next();
+      });
+    }
+  }
+
+  static create<R extends Router<any, any>>(
+    config: OneSchemaRouterConfig<R>,
+  ): OneSchemaRouter<{}, R> {
+    return new OneSchemaRouter({}, config);
+  }
+
+  declare<
+    Route extends RoughRoute,
+    Name extends string,
+    Endpoint extends RouterEndpointDefinition<Name>,
+  >(
+    endpoint: Endpoint & { route: Route },
+  ): OneSchemaRouter<Schema & { [route in Route]: Endpoint }, R> {
+    // @ts-expect-error
+    this.schema[endpoint.route] = endpoint;
+
+    return this as any;
+  }
+
+  implement<Route extends keyof Schema & string>(
+    route: Route,
+    implementation: EndpointImplementation<
+      Route,
+      z.infer<Schema[Route]['request']>,
+      z.infer<Schema[Route]['response']>,
+      R
+    >,
+  ) {
+    const endpoint = this.schema[route];
+
+    implementRoute(
+      route,
+      this.router,
+      (ctx, data) => {
+        const res = endpoint.request.safeParse(data);
+        if (!res.success) {
+          const friendlyError = fromZodError(res.error, {
+            prefix: 'The request input did not conform to the required schema',
+          });
+          return ctx.throw(400, friendlyError.message);
+        }
+        return res.data;
+      },
+      implementation,
+    );
+
+    return this;
+  }
+
+  middleware(): Router.Middleware {
+    return compose([this.router.routes(), this.router.allowedMethods()]);
+  }
+}
+
+const convertRouterSchemaToJSONSchemaStyle = <Schema extends ZodSchema>(
+  schema: Schema,
+): OneSchemaDefinition => {
+  const oneSchema: OneSchemaDefinition = { Endpoints: {} };
+
+  for (const [endpoint, definition] of Object.entries(schema)) {
+    oneSchema.Endpoints[endpoint] = {
+      Name: definition.name,
+      Description: definition.description,
+      // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
+      Request: zodToJsonSchema(definition.request) as any,
```

Comment:
I see we used `any` in different places here. I'm assuming getting full type safety with generics was a bit tedious. Maybe it's worth commenting on why that's the case just so we know that `any` was an acceptable solution here.

Reply:
> I'm assuming getting full type safety with generics was a bit tedious.

This. I will make a TODO to follow-up with inline comments.